keps/sig-storage/2324-volume-snapshot-namespace-transfer/README.md

```diff
@@ -0,0 +1,485 @@
+
+## Summary
+
+Currently, VolumeSnapshot resources are tied to a specific namespace. This KEP proposes a method for transferring
+these resources between namespaces, allowing users to conveniently move these resources without manual creation
+of each independent object (VolumeSnapshot, VolumeSnapshotContent).
+
+This extends the [API for PVC Namespace Transfer](https://github.com/kubernetes/enhancements/pull/2110/files) to also apply for VolumeSnapshots.
+
+## Motivation
+
+Allow Kubernetes users to quickly and conveniently transfer VolumeSnapshots across namespaces.
+
+### Goals 
+
+- Extend the API proposed for PVC Namespace Transfer to apply for VolumeSnapshots
+
+### Non-Goals
+
+## Proposal
+
+We are going to allow users to transfer VolumeSnapshot resources between namespaces, provided
+that both the source and target namespace have a user that has agreed to the transfer.
+
+The source and target VolumeSnapshot will point to the same VolumeSnapshotContent, which
+references the backing snapshot on the storage backend.
+
+Once the transfer is completed, the source VolumeSnapshot should be deleted, so that only
+a single VolumeSnapshot stays mapped to a single VolumeSnapshotContent.
+
+### User Stories
+
+Users should be able to transfer VolumeSnapshots that reference the same VolumeSnapshotContent
+across namespaces. This allows users to quickly share snapshots, utilizing data for testing
+or backups in separate namespaces.
+
+#### Story 1
+
+The `prod` namespace contains a PVC that holds user information. VolumeSnapshots are taken of 
+this PVC on a regular basis.
+
+An existing VolumeSnapshot can be transferred to a test namespace. Once transferred, a PVC
+can be created from the snapshot for testing purposes.  
+
+### Risks and Mitigations
+
+- As long as we reference the existing VolumeSnapshotContent, there should be no issues with encrypted VolumeSnapshotContents. The secret reference is stored on the VolumeSnapshotContent and VolumeSnapshotClass, not the VolumeSnapshot itself. If we take a different approach, such as re-creating the VolumeSnapshotContent, then we will need to ensure these credentials are propagated to the new content.
+- The VolumeSnapshot must be `readyToUse` before the transfer begins.
+- Any owners of the destination namespace must accept the transfer. By using the StorageTransfer API, this ensures an owner in the destination namespace is aware of the transfer.
+
+## Design Details
+
+Refer to [API for PVC Namespace Transfer](https://github.com/kubernetes/enhancements/pull/2110/) for 
+details concerning the StorageTransfer API.
+
+A `StorageTransferRequest` resource will be created in the source namespace:
+
+```yaml
+apiVersion: v1alpha1
+kind: StorageTransferRequest
+metadata:
+  name: transfer-foo
+  namespace: target-namespace
+spec:
+  source:
+    name: foo
+    namespace: source-namespace
+    kind: VolumeSnapshot
+  approvalName: approve-foo
```

Comment:
How does a transfer request know `approvalName`? does this gets filled later on by snapshot controller?

Reply:
IMO it's part of the agreement. User who gives the snapshots must pass this name to the user who accepts the snapshots, outside of Kubernetes API.

Reply:
BTW, I'm checking OpenStack, which has very similar concept called [volume transfer between projects](https://docs.openstack.org/cinder/latest/cli/cli-manage-volumes.html#transfer-a-volume), and here OpenStack actually generates a token, that the current owner of the volume (old project) must give to the new owner (new project).

The KEP should note that the name should be at least something that the users can't guess. Using a random string in the example would be helpful.